src/policy/v3_policy.cpp

```diff
@@ -0,0 +1,111 @@
+    std::unordered_map<Txid, Wtxid, SaltedTxidHasher> non_v3_txid_to_wtxid;
+    for (const auto& tx : package) {
+        if (tx->nVersion == 3) {
+            v3_txid_to_wtxid.emplace(tx->GetHash(), tx->GetWitnessHash());
+        } else {
+            non_v3_txid_to_wtxid.emplace(tx->GetHash(), tx->GetWitnessHash());
+        }
+    }
+    for (const auto& tx : package) {
+        if (tx->nVersion == 3) {
+            for (const auto& input : tx->vin) {
+                if (auto it = non_v3_txid_to_wtxid.find(input.prevout.hash); it != non_v3_txid_to_wtxid.end()) {
+                    return std::make_tuple(it->second, tx->GetWitnessHash(), true);
+                }
+            }
+        } else {
+            for (const auto& input : tx->vin) {
+                if (auto it = v3_txid_to_wtxid.find(input.prevout.hash); it != v3_txid_to_wtxid.end()) {
+                    return std::make_tuple(it->second, tx->GetWitnessHash(), false);
+                }
+            }
+        }
+    }
+    return std::nullopt;
+}
+
+std::optional<std::string> CheckV3Inheritance(const CTransactionRef& ptx,
+                                              const CTxMemPool::setEntries& ancestors)
+{
+    for (const auto& entry : ancestors) {
+        if (ptx->nVersion != 3 && entry->GetTx().nVersion == 3) {
+            return strprintf("tx that spends from %s must be nVersion=3",
+                             entry->GetTx().GetWitnessHash().ToString());
+        } else if (ptx->nVersion == 3 && entry->GetTx().nVersion != 3) {
+            return strprintf("v3 tx cannot spend from %s which is not nVersion=3",
+                             entry->GetTx().GetWitnessHash().ToString());
+        }
+    }
+    return std::nullopt;
+}
+
+std::optional<std::string> ApplyV3Rules(const CTransactionRef& ptx,
+                                        const CTxMemPool::setEntries& ancestors,
+                                        const std::set<Txid>& direct_conflicts,
+                                        int64_t sigops)
+{
+    // These rules only apply to transactions with nVersion=3.
+    if (ptx->nVersion != 3) return std::nullopt;
+
+    if (ancestors.size() + 1 > V3_ANCESTOR_LIMIT) {
+        return strprintf("tx %s would have too many ancestors", ptx->GetWitnessHash().ToString());
+    }
+    if (ancestors.empty()) {
+        return std::nullopt;
+    } else {
+        const auto tx_weight{GetTransactionWeight(*ptx)};
+        // If this transaction spends V3 parents, it cannot be too large.
+        if (tx_weight > V3_CHILD_MAX_WEIGHT) {
+            return strprintf("v3 child tx is too big: %u > %u weight units", tx_weight, V3_CHILD_MAX_WEIGHT);
+        }
+        if (sigops > V3_CHILD_MAX_SIGOPS) {
+            return strprintf("v3 child tx has too many sigops: %u > %u sigops", sigops, V3_CHILD_MAX_SIGOPS);
+        }
+        // Any ancestor of a V3 transaction must also be V3.
+        const auto& parent_entry = *ancestors.begin();
```

Comment:
This logic only works if `V3_ANCESTOR_LIMIT <= 2`, I think?  Having a constant would imply that you could consider changing the value, but if the logic breaks with higher values then I think we should `static_assert` that the constant is in the range we expect.